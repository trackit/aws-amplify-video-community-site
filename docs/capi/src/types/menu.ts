import {PageLink} from "./page";

export interface WrittenMenu {
  title?: string;
  items: string[];
}

export interface MenuGroup {
  title: string;
  items: PageLink[];
}

export type Menu = MenuGroup[];
