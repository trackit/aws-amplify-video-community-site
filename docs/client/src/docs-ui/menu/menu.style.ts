import {css} from "emotion";

export const menuStyle = css`
  display: block;
  padding: 0 2.5rem;
`;

export const menuItemContainerStyle = css`
  display: flex;
  flex-direction: column;
`;

export const menuBreakStyle = css`
  border-top: 0.0625rem solid var(--color-grey-lt);
  margin: 1.5rem -2.5rem 1.5rem -2.5rem;
`;

export const productRootLink = css`
  display: flex;
  flex-direction: row;
  font-weight: 700;
  margin-top: 2rem;

  > stencil-route-link,
  > stencil-route-link > a {
    display: block;
    width: 100%;
  }

  > stencil-route-link > a {
    padding: 0.5rem;
  }

  &:hover {
    background-color: var(--bg-color-hover);
  }

  a {
    color: var(--color-dark-hv);
  }

  img {
    width: 1rem;
    margin-right: 0.5rem;
  }
`;

export const verticalGapStyle = css`
  width: 100%;
  height: 2rem;
`;

export const activeLinkStyle = css`
  background-color: var(--bg-color-hover);
`;

export const discordLinkStyle = css`
  > a {
    display: flex;
    background-color: var(--color-discord-blue);
    flex-direction: row;
    align-items: center;
    justify-content: center;
    border-radius: 0.125rem;
    padding: 0.675rem 1rem;
    opacity: 0.9;

    &:hover {
      opacity: 1;
    }
  }

  a {
    display: flex;
    color: var(--color-white);
    font-weight: 700;
  }

  img {
    width: 1.5rem;
    height: 1.5rem;
    margin-right: 0.25rem;
  }
`;
