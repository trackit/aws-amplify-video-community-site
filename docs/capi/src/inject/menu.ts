import * as t from "../types";
import * as path from "path";
import {createPageLink} from "../utils";

/**
 * iterates through all pages who should have a menu, stitches together
 * their menus, and injects them in place into the page obj
 */
export const injectMenu = (ctx: t.Ctx): void => {
  // for every product directory
  ctx.productDirs.forEach((productDir) => {
    // get the pages to whom the menu applies
    const menuForPaths = ctx.pageSrcPaths.filter((path) =>
      path.startsWith(productDir),
    );

    // get the written menu object or throw an error if it doesn't exist
    const productDirMenuPath = path.join(productDir, "menu.json");
    const productDirMenu = ctx.menuBySrcPath.get(productDirMenuPath);
    if (!productDirMenu) {
      throw new Error(`Create a "${productDirMenuPath}" file.`);
    }

    // construct the menu, ... or throw errors ;)
    const menu: t.Menu = productDirMenu.items.map((item) => {
      const nestedMenuPath = path.join(productDir, item, "menu.json");
      const nestedMenu = ctx.menuBySrcPath.get(nestedMenuPath);

      if (!nestedMenu) {
        throw new Error(`Create a "${nestedMenuPath}" file.`);
      }

      if (!nestedMenu.title) {
        throw new Error(`"title" field is required in "${nestedMenuPath}"`);
      }

      return {
        title: nestedMenu.title,
        items: nestedMenu.items.map((mdFileName) => {
          // construct the `srcPath`, which we'll use to get the corresponding page obj
          const pagePath = [
            path.join(productDir, item, mdFileName),
            ".md",
          ].join("");

          // retrieve the page obj
          const page = ctx.pageBySrcPath.get(pagePath);
          if (!page) {
            throw new Error(
              `Ran into problems resolving "${pagePath}", referenced from "${nestedMenuPath}"`,
            );
          }
          page.sectionTitle = nestedMenu.title;

          // create and return the page link
          return createPageLink(page);
        }),
      };
    });

    // get the product root page link
    const productDirFolderName = productDir.split(path.sep).pop();
    const productRootPath = path.join(productDir, `${productDirFolderName}.md`);
    const productRootPage = ctx.pageBySrcPath.get(productRootPath);
    if (!productRootPage) {
      throw new Error(`Create a "${productRootPath}" file.`);
    }
    const productRootPageLink = createPageLink(productRootPage);

    // add the menu to the product subpages, ... or throw an error ;)
    menuForPaths.forEach((pagePath) => {
      const page = ctx.pageBySrcPath.get(pagePath);
      if (!page) {
        throw new Error(`Ran into problems resolving "${pagePath}".`);
      }
      page.menu = menu;
      page.productRootLink = productRootPageLink;
    });
  });
};
