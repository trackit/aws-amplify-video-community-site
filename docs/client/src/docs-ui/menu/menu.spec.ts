import {DocsMenu} from "./menu";
import {newSpecPage} from "@stencil/core/testing";

describe("docs-menu", () => {
  describe("Component logic", () => {
    let menu: DocsMenu;
    beforeEach(() => (menu = new DocsMenu()));

    it("should init `page` as `undefined`", () =>
      expect(menu.page).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [DocsMenu],
            html: `<docs-menu />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
