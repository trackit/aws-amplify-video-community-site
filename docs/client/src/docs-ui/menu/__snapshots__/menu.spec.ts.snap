// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs-menu Render logic should render 1`] = `<docs-menu></docs-menu>`;
